pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -18,158 +18,219 @@ package credentials
+	if err != nil {
+		return cfg, err
+	}
+	params := &ecr.GetAuthorizationTokenInput{RegistryIds: []*string{aws.String(parsed.registryID)}}
+	output, err := getter.GetAuthorizationToken(params)
+	if err != nil {
+		return cfg, err
+	}
+	if output == nil {
+		return cfg, errors.New("unable to get ECR token")
+	}
+	if len(output.AuthorizationData) == 0 {
+		return cfg, errors.New("unable to get authorization data")
+	}
+	data := output.AuthorizationData[0]
+	if data.AuthorizationToken == nil {
+		return cfg, errors.New("unable to get authorization token")
+	}
+	if data.ProxyEndpoint == nil {
+		return cfg, errors.New("unable to get proxy endpoint")
+	}
 
-// registryURL has different suffix in AWS China region
-func registryURL(region string) string {
-	dnsSuffix := awsStandardDNSSuffix
-	// deal with aws none standard regions
-	if strings.HasPrefix(region, awsChinaRegionPrefix) {
-		dnsSuffix = awsChinaDNSSuffix
+	decodedToken, err := base64.StdEncoding.DecodeString(aws.StringValue(data.AuthorizationToken))
+	if err != nil {
+		return cfg, err
+	}
+	parts := strings.SplitN(string(decodedToken), ":", 2)
+	user := parts[0]
+	password := parts[1]
+	creds := credentialprovider.DockerConfigEntry{
+		Username: user,
+		Password: password,
+		// ECR doesn't care and Docker is about to obsolete it
+		Email: "not@val.id",
 	}
-	return fmt.Sprintf(registryURLTemplate, region, dnsSuffix)
+	entry := &cacheEntry{
+		expiresAt:   *data.ExpiresAt,
+		credentials: creds,
+		registry:    parsed.registry,
+	}
+	if err := p.cache.Add(entry); err != nil {
+		return cfg, err
+	}
+	cfg[entry.registry] = entry.credentials
+	return cfg, nil
 }
 
-// RegisterCredentialsProvider registers a credential provider for the specified region.
-// It creates a lazy provider for each AWS region, in order to support
-// cross-region ECR access. They have to be lazy because it's unlikely, but not
-// impossible, that we'll use more than one.
-// This should be called only if using the AWS cloud provider.
-// This way, we avoid timeouts waiting for a non-existent provider.
-func RegisterCredentialsProvider(region string) {
-	klog.V(4).Infof("registering credentials provider for AWS region %q", region)
+type parsedURL struct {
+	registryID string
+	region     string
+	registry   string
+}
 
-	credentialprovider.RegisterCredentialProvider("aws-ecr-"+region,
-		&lazyEcrProvider{
-			region:    region,
-			regionURL: registryURL(region),
-		})
+// parseRepoURL parses and splits the registry URL into the registry ID,
+// region, and registry.
+// <registryID>.dkr.ecr.<region>.amazonaws.com
+// <registryID>.dkr.ecr.<region>.amazonaws.com.cn
+func parseRepoURL(repoToPull string) (*parsedURL, error) {
+	parsed, err := url.Parse("https://" + repoToPull)
+	if err != nil {
+		klog.Errorf("unable to parse repository URL %v", err)
+		return nil, err
+	}
+	splitURL := strings.Split(parsed.Host, ".")
+	if len(splitURL) < 4 || splitURL[2] != "ecr" || splitURL[4] != "amazonaws" {
```

Comment:
`splitURL[2]` can also (validly) be `ecr-fips`.  If it's `ecr-fips`, you should also override the ECR API endpoint.  You can construct the correct endpoint with something like this:

```go
resolver := endpoints.DefaultResolver()
endpoint, err := resolver.EndpointFor("ecr-fips", region, func(opts *endpoints.Options) {
		opts.ResolveUnknownService = true
})
```

and then you can pass the endpoint in when creating the API client.

Reply:
Will make this as a separate PR, thanks!